# DSCI 100 Project: Final Report

**Authors**: 

**Predicting Usage of a Video Game Research Server**
A research group in the Department of Computer Science at UBC (PLAI) has a goal to aid in advancing the limits of artificial intelligence. They host a Minecraft server that records players' actions as they navigate through the world. Since running the project is so complex, they want to recruit players who will contribute large amounts of gameplay data, as well as make sure they have the resources to handle the amount of players they attract. In an attempt to optimize their players' contributions, they want to understand what demographics of players will have the greatest gameplay time. This project answers their question of how player demographics relate to total gameplay time. 

## The Question
**Question 2**: Which kinds of players are most likely to contribute a large amount of data, and how can we identify them for targeted recruitment?

**Predictive Question** Can we predict the total `played_hours` from the `players.csv` dataset based on the two predictors, `experience` and `subscribe`.

**Response variable**:

`played_hours` - total play time and data contribution

**Explanatory variables**:

`experience` - player experience level

`subscribe` -  subscription status of player

The dataset allows us to examine whether `experience` or `subscribe` relates to total playtime. If certain experience levels or subscribtion status tend to play more, those groups may contribute more gameplay data. This helps identify the player demographics most likely to provide substantial data for the research team.


## Data Description: The Players
The dataset used in this analysis is a player demographics and engagement dataset that contains 196 observations (rows) and 9 variables (columns). The dataset records information about individual players, their gaming experience level, subscription status, demographic details, and the amount of time they have spent playing.

- Number of observations: 196 (each row represents one unique player)
- Number of variables: 9
- Observational unit: Player-level data
- Purpose: Identify which demographic characteristics correspond to higher gameplay time (and therefore more data contributed)
  
**Variables**
- `played_hours` (numeric) - Total hours each player spent on the server
- `experience` (categorical) -  Self-reported experience level in Minecraft
- `gender` (categorical) - Player’s gender
- `subscribe` (Boolean) - Subscription status of player
- `hashedEmail` (String) - The players email, hashed for privacy
- `name` (String) - Name of player
- `age` (Numeric) - Age of player
- `individualID` (String) - Unique ID of player
- `organizationName` (String) - Organization name

**Issues to consider**
- Missing identifiers (`individualId`, `organizationName`).
- `hashedEmail` is anonymous for privacy.
- `played_hours` may include idle time, which could overestimate actual engagement.

Source of the original dataset:
https://drive.google.com/file/d/1Mw9vW0hjTJwRWx0bDXiSpYsO3gKogaPz/edit


In [2]:
# Import libraries
import pandas as pd
import altair as alt
# Read the datasets
url= "https://raw.githubusercontent.com/sydlpeters/dsci-group-2025w1-group-101-1/refs/heads/main/data/players.csv"
# Load the dataset 
players = pd.read_csv(url)
players

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [3]:
# Drop columns that are entirely missing(NaN)
players_tidy = players.drop(columns=["individualId", "organizationName"])
# Preview tidy players_tidy
players_tidy.head()

experience  subscribe                                        hashedEmail  \
0        Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1    Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2    Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3    Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4    Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   

   played_hours       name  gender  age  
0          30.3     Morgan    Male    9  
1           3.8  Christian    Male   17  
2           0.0      Blake    Male   17  
3           0.7      Flora  Female   21  
4           0.1      Kylie    Male   21

In [4]:
players_tidy.describe()

played_hours         age
count    196.000000  196.000000
mean       5.845918   21.280612
std       28.357343    9.706346
min        0.000000    8.000000
25%        0.000000   17.000000
50%        0.100000   19.000000
75%        0.600000   22.000000
max      223.100000   99.000000

**Summary Statistics**

Most players recorded very little playtime, with three-quarters spending under an hour in total (Q3 = 0.6 hours). Only a small number logged substantially more time, which raises the overall average (mean = 5.85 hours; median = 0.1 hours) and affects the dataset (STD = 28.36 hours). **Most players played very little (close to 0 hours), but a few played way more, making the dataset highly uneven.**

The player population is primarily young adults, concentrated between the late teens and early twenties (mean = 21.28; Q1–Q3 = 17–22 years). 

In [5]:
# Experienced comapred with Average played hours played
plot1= alt.Chart(players_tidy).mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("mean(played_hours):Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot1

alt.Chart(...)

In [17]:
# Experienced comapred with Average played hours played
plot1= alt.Chart(players_tidy).mark_bar().encode(
    x=alt.X("experience:N").title("Experience Level"),
    y=alt.Y("played_hours:Q").title("Average hours Played(hr)"),
    color = alt.Color("experience:N").title("Experience Level")
)
plot1

alt.Chart(...)

In [6]:
# Subscription status comapred with Average hours played
plot2= alt.Chart(players_tidy).mark_bar().encode(
    x=alt.X("subscribe:N").title("Subscription Status"),
    y=alt.Y("mean(played_hours)").title("Average Hours Played(hr)"),
    color = alt.Color("subscribe:N").title("Subscription Status")
)
plot2

alt.Chart(...)

In [7]:
experience_map = {
    "Beginner" : 0,
    "Regular": 1,
    "Amateur": 2,
    "Veteran": 3,
    "Pro": 4
}

players["experience"] = players["experience"].map(experience_map)
players

experience  subscribe                                        hashedEmail  \
0             4       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1             3       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2             3      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3             2       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4             1       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..          ...        ...                                                ...   
191           2       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192           3      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193           2      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194           2      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195           4       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [8]:
subscribe_map = {
    False : 0,
    True : 1,
}

players["subscribe"] = players["subscribe"].map(subscribe_map)
players

experience  subscribe                                        hashedEmail  \
0             4          1  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1             3          1  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2             3          0  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3             2          1  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4             1          1  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..          ...        ...                                                ...   
191           2          1  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192           3          0  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193           2          0  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194           2          0  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195           4          1  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [9]:
# import the K-NN regression model
from sklearn.model_selection import GridSearchCV, cross_validate, train_test_split
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

# import the K-NN regression model
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error



# split data!
players_training, players_testing = train_test_split(
    players, test_size=0.2, random_state=1234
)

# set target and predictors
players_x_train = players_training[["experience"]]
players_y_train = players_training["played_hours"]

players_x_test = players_testing[["experience"]]
players_y_test = players_testing["played_hours"]

players_training

experience  subscribe                                        hashedEmail  \
114           0          1  ae8d8a9dcf80b38466f89466201e6d594eb94d6994eda1...   
187           2          1  e3f0ad9aadd27f3d1d9197e58546d045018daa76767503...   
39            2          1  6ef84a204d64edc3b3779127b13f298d0c70f96568486f...   
25            1          1  5baba1651a0b92788bc0d6dcdf00be64af1cf9f0015bbe...   
131           2          0  11bf6125c4264b3a8f3bffa57b33bd598e2ea1ecd6331a...   
..          ...        ...                                                ...   
152           0          1  aea049eaa7cb10db386a62990220d205ceb2a4c473cae3...   
116           3          0  58893f3187db90bf0690c88f06e6aa5f8ea9ff9691ca8e...   
53            2          1  fc0224c81384770e93ca717f32713960144bf0b52ff676...   
38            3          1  d782933acd14c834e53dea816005a3583cb87710f7347a...   
47            1          0  453774822d52d6891593948224570a750c4334616ff8aa...   

     played_hours     name             gender  age  individualId  \
114           1.0     Ella             Female   17           NaN   
187           0.0   Jasper               Male   17           NaN   
39            0.0   Vivian               Male   17           NaN   
25            0.6  Kendall             Female   28           NaN   
131           0.0   Olivia             Female   23           NaN   
..            ...      ...                ...  ...           ...   
152           0.2    Aarav  Prefer not to say   17           NaN   
116           0.0     Noah  Prefer not to say   20           NaN   
53            0.2    Gemna               Male   27           NaN   
38            0.0   Ishaan               Male   17           NaN   
47            0.0   Edmund  Prefer not to say   23           NaN   

     organizationName  
114               NaN  
187               NaN  
39                NaN  
25                NaN  
131               NaN  
..                ...  
152               NaN  
116               NaN  
53                NaN  
38                NaN  
47                NaN  

[156 rows x 9 columns]

In [10]:
# preprocess the data, make the pipeline
players_preprocessor = make_column_transformer(
    (
        StandardScaler(),
        [
            "experience",
            ],
    )
)
players_pipeline = make_pipeline(players_preprocessor, KNeighborsRegressor())

players_cv = pd.DataFrame(
    cross_validate(
        players_pipeline,
        players_x_train,
        players_y_train,
        cv=5,
        scoring="neg_root_mean_squared_error",
        return_train_score=True,
    )
)
players_cv

fit_time  score_time  test_score  train_score
0  0.004783    0.002978   -9.099486   -21.719026
1  0.001159    0.000000   -5.920221   -22.051516
2  0.000000    0.007003  -28.129594   -16.957817
3  0.001758    0.001875   -1.713133   -22.352437
4  0.002027    0.001726  -32.491334   -15.254916

In [11]:
# create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 50, 1),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)

# fit the GridSearchCV object
players_result = pd.DataFrame(players_tuned.fit(players_x_train, players_y_train).cv_results_)

players_result

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003430      0.003360         0.000495        0.000616   
1        0.001599      0.002038         0.001629        0.003257   
2        0.003740      0.004854         0.000000        0.000000   
3        0.001484      0.002112         0.002314        0.002981   
4        0.001788      0.002422         0.001806        0.003611   
5        0.003259      0.005277         0.000101        0.000203   
6        0.000308      0.000615         0.003038        0.004601   
7        0.000201      0.000402         0.003132        0.004529   
8        0.002347      0.004694         0.001008        0.002015   
9        0.002214      0.004428         0.001122        0.002243   
10       0.001224      0.002448         0.002350        0.003111   
11       0.002351      0.000465         0.000652        0.000833   
12       0.002871      0.005743         0.000175        0.000349   
13       0.003137      0.006274         0.000000        0.000000   
14       0.003509      0.003316         0.000830        0.001163   
15       0.001603      0.001697         0.001294        0.001070   
16       0.002048      0.002570         0.000075        0.000150   
17       0.003557      0.003954         0.000000        0.000000   
18       0.003637      0.005175         0.000604        0.000804   
19       0.001266      0.001040         0.001063        0.000921   
20       0.001674      0.000847         0.001108        0.000921   
21       0.002025      0.000196         0.001103        0.000919   
22       0.002118      0.002814         0.000591        0.000783   
23       0.002067      0.000678         0.001193        0.000599   
24       0.000537      0.000668         0.002697        0.002433   
25       0.002887      0.003709         0.000376        0.000753   
26       0.000567      0.001134         0.000000        0.000000   
27       0.001753      0.003506         0.002957        0.005299   
28       0.000061      0.000121         0.003270        0.006540   
29       0.002577      0.004668         0.001045        0.001336   
30       0.001855      0.000303         0.000806        0.000987   
31       0.000084      0.000167         0.002480        0.004960   
32       0.002903      0.004581         0.000415        0.000830   
33       0.003351      0.002876         0.001047        0.001555   
34       0.001540      0.000780         0.000750        0.000922   
35       0.001812      0.003145         0.001652        0.003305   
36       0.003135      0.005632         0.000276        0.000552   
37       0.001198      0.000978         0.003456        0.004874   
38       0.001553      0.000783         0.001641        0.001038   
39       0.001428      0.001749         0.000072        0.000145   
40       0.003015      0.004062         0.000201        0.000403   
41       0.002201      0.003768         0.000000        0.000000   
42       0.001063      0.000905         0.002417        0.001984   
43       0.000596      0.001191         0.001759        0.003519   
44       0.000414      0.000678         0.001495        0.002346   
45       0.002918      0.005836         0.000445        0.000889   
46       0.000625      0.001251         0.002909        0.005330   
47       0.003133      0.004534         0.000000        0.000000   
48       0.001221      0.001624         0.000289        0.000578   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                                      11   
11                                      12   
12   

In [12]:
# Retrieve the CV scores


# get the best parameter values
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 9}

In [13]:
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(15.027051604673355)

In [14]:
knn = KNeighborsRegressor(n_neighbors = 10)
player_pipe = make_pipeline(players_preprocessor, knn)
player_pipe.fit(players_x_train, players_y_train)

players_predictions = players_testing.assign(prediction = player_pipe.predict(players_x_test))
players_predictions

experience  subscribe                                        hashedEmail  \
101           2          1  25879aecc205544bc6505f9faf768356e0a3b712605730...   
51            1          1  b622593d2ef8b337dc554acb307d04a88114f2bf453b18...   
146           4          1  5669c0f4b50dbfe3e2f851e4bb89fa43c55cd1f71ba362...   
153           0          1  0ce7bfa910d47fc91f21a7b3acd8f33bde6db57912ce02...   
106           1          0  63139b524e44d156daf265875fb8078bb2b24d6b7520a3...   
59            3          1  ca20f724571080b997e0efa874b9611e9f280c1af5f68f...   
161           3          0  f174555ae3ff613d5d0b98f83f26f689619d92ef272f28...   
167           0          0  42eafe96ed5c1684e3b5cc614d1b01a117173d3ec6898a...   
193           2          0  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
88            0          1  a1094a440899e69e804f888f44d1154ec0b7675d05d977...   
29            3          0  951e54f7376e2b2f0915e9e3646c701af4a2fe839385b1...   
130           2          1  fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33...   
66            3          0  70db3a16ef58cf8bf6d8658c426ba4a5fcd7523a0abdd4...   
74            1          1  bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431...   
48            3          1  b3510c708bd50bf9f75e6e02bb6fe14edb705e0ea671ee...   
100           2          1  26362473ca8fab46358dbab9bcbcf7114bbcbadef90e6b...   
123           0          0  e74c60a92c0100e7240be56d66969db85856152b048c63...   
31            3          1  f1b432523542f90c61176a555ccb2144468d76c91a32d7...   
95            1          0  dea781ecb85a9e85c8d1fa62b0c2a98c77a51f5933c104...   
137           3          1  d711ff0c9d03664676d1ce80ca7ac5e57f7ebbc0d04bd4...   
6             1          1  8e594b8953193b26f498db95a508b03c6fe1c24bb5251d...   
54            0          0  5e5c25a773be7a62638a163d773534e575a5ad57821047...   
97            3          0  d3ca24e4d7fe8ffe2a821ebd3b841252950ca53e4d659a...   
5             2          1  f58aad5996a435f16b0284a3b267f973f9af99e7a89bee...   
110           2          1  827f9a82b47cdf21098bc7bb1d4241550dc6146ba9ae8b...   
91            0          1  eeb961422a3f9841018d4eb729ddc40ef103cc1205577d...   
134           0          1  fef4e1bed8c3f6dcd7bcd39ab21bd402386155b2ff8c8e...   
180           2          0  93efabc9f7de58f44c457bac715cf8f9f44cd88cc574fc...   
7             2          0  1d2371d8a35c8831034b25bda8764539ab7db0f6393869...   
168           3          1  9ba67e127e4e70d909f6237644bb4674b17bc6ae5c9e73...   
94            0          1  e3175c379bbb83c8786d2dd719cf844575aca24d2c81d6...   
9             3          1  bbe2d83de678f519c4b3daa7265e683b4fe2d814077f90...   
58            1          1  f2826fb8dbce4d450348f99cb27ade184b713998d96797...   
92            4          1  fcab03c6d3079521e7f9665caed0f31fe3dae6b5ccb86e...   
111           1          1  336adf9b9070abb44eaec22a48b04d8cda6cf0c98a46ee...   
43            3          1  126ec350278422ba60ad575d230c19522fd568094b7c0f...   
77            1          1  e21a324ccf5c873bafe82e47d5137b36aa312ee4803eeb...   
142           0          1  6ca8a90e4f5c61bf36baf556d446873c4e894e7a62f919...   
119           0          1  d7e7a5ff16f94d1e793490b521671f7ca56d77820df9b9...   
40            1          1  0d4d71be33e2bc7266ee4983002bd930f69d304288a866...   

     played_hours       name             gender  age  individualId  \
101           0.0        Eli             Female   17           NaN   
51          218.1       Akio         Non-binary   20           NaN   
146           0.0      Padma         Non-binary   25           NaN   
153           0.1     Osiris               Male   17           NaN   
106           0.0      Elias               Male   23           NaN   
59            0.2     Edward               Male   38           NaN   
161           0.0     Finley         Non-binary   17           NaN   
167           0.3     Ariana             Female   17           NaN   
193           0.0      Dylan  Prefer not to say   17 

In [15]:

players_plot = (alt.Chart(players_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "K=39"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction"
))
players_plot

alt.LayerChart(...)

In [ ]:
players_predictions = players_predictions.rename(columns={"prediction":"prediction_80"})

players_cv = pd.DataFrame(
    cross_validate(
        players_pipeline,
        players_x_train,
        players_y_train,
        cv=5,
        scoring="neg_root_mean_squared_error",
        return_train_score=True,
    )
)

# create the 5-fold GridSearchCV object
param_grid = {
    "kneighborsregressor__n_neighbors": range(1, 50, 1),
}

players_tuned = GridSearchCV(
    estimator=players_pipeline,
    param_grid=param_grid,
    cv=5,
    scoring="neg_root_mean_squared_error",
)



players_result = pd.DataFrame(players_tuned.fit(players[["experience"]], players[["played_hours"]]).cv_results_)

players_result

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.002222      0.002990         0.000811        0.001051   
1        0.002604      0.003052         0.003820        0.005448   
2        0.001646      0.001408         0.002560        0.004346   
3        0.001826      0.003652         0.001066        0.001328   
4        0.005662      0.005685         0.000789        0.000970   
5        0.002686      0.004481         0.000366        0.000733   
6        0.002885      0.004464         0.003337        0.004505   
7        0.002628      0.002766         0.000596        0.001192   
8        0.000204      0.000407         0.002823        0.004320   
9        0.005559      0.006070         0.000399        0.000797   
10       0.002847      0.005695         0.000112        0.000225   
11       0.003112      0.005574         0.000526        0.001052   
12       0.002831      0.004670         0.000241        0.000482   
13       0.000624      0.000767         0.004506        0.005442   
14       0.003595      0.004267         0.000000        0.000000   
15       0.002437      0.003743         0.001244        0.001916   
16       0.002676      0.005352         0.000271        0.000542   
17       0.003055      0.006110         0.000317        0.000633   
18       0.002605      0.005209         0.000400        0.000800   
19       0.004404      0.005024         0.001043        0.001450   
20       0.002890      0.004108         0.001232        0.001860   
21       0.000270      0.000539         0.002510        0.003697   
22       0.002016      0.004031         0.000937        0.001298   
23       0.000962      0.001245         0.003121        0.004197   
24       0.000481      0.000600         0.002390        0.003884   
25       0.000554      0.001109         0.004648        0.005256   
26       0.002035      0.003439         0.001989        0.001798   
27       0.001003      0.002005         0.002335        0.004670   
28       0.001215      0.002431         0.002403        0.004158   
29       0.001477      0.001275         0.001829        0.002964   
30       0.005084      0.005382         0.000552        0.000727   
31       0.002573      0.001875         0.001938        0.001124   
32       0.000290      0.000581         0.003676        0.002696   
33       0.002540      0.001402         0.001661        0.001099   
34       0.002096      0.001363         0.000868        0.000741   
35       0.002545      0.003353         0.000000        0.000000   
36       0.002632      0.003348         0.000408        0.000817   
37       0.004725      0.004035         0.000092        0.000185   
38       0.003602      0.003939         0.000894        0.000751   
39       0.001028      0.000866         0.002338        0.003867   
40       0.000468      0.000936         0.002873        0.005103   
41       0.000000      0.000000         0.003071        0.004658   
42       0.000268      0.000535         0.005081        0.004735   
43       0.001054      0.002108         0.000000        0.000000   
44       0.001234      0.002468         0.003335        0.004165   
45       0.000401      0.000801         0.002729        0.004816   
46       0.001354      0.002417         0.000000        0.000000   
47       0.002037      0.003441         0.003060        0.006120   
48       0.001225      0.001037         0.002803        0.001536   

    param_kneighborsregressor__n_neighbors  \
0                                        1   
1                                        2   
2                                        3   
3                                        4   
4                                        5   
5                                        6   
6                                        7   
7                                        8   
8                                        9   
9                                       10   
10                                      11   
11                                      12   
12   

In [21]:
players_min = players_tuned.best_params_
players_min

{'kneighborsregressor__n_neighbors': 12}

In [22]:
players_best_RMSPE = -players_tuned.best_score_
players_best_RMSPE

np.float64(24.622432143663325)

In [30]:
knn = KNeighborsRegressor(n_neighbors = 12)
player_pipe = make_pipeline(players_preprocessor, knn)
player_pipe.fit(players[["experience"]], players[["played_hours"]])

print(players_predictions)

players_predictions = players.assign(prediction_all = player_pipe.predict(players[["experience"]]))
players_predictions

     experience  subscribe                                        hashedEmail  \
101           2          1  25879aecc205544bc6505f9faf768356e0a3b712605730...   
51            1          1  b622593d2ef8b337dc554acb307d04a88114f2bf453b18...   
146           4          1  5669c0f4b50dbfe3e2f851e4bb89fa43c55cd1f71ba362...   
153           0          1  0ce7bfa910d47fc91f21a7b3acd8f33bde6db57912ce02...   
106           1          0  63139b524e44d156daf265875fb8078bb2b24d6b7520a3...   
59            3          1  ca20f724571080b997e0efa874b9611e9f280c1af5f68f...   
161           3          0  f174555ae3ff613d5d0b98f83f26f689619d92ef272f28...   
167           0          0  42eafe96ed5c1684e3b5cc614d1b01a117173d3ec6898a...   
193           2          0  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
88            0          1  a1094a440899e69e804f888f44d1154ec0b7675d05d977...   
29            3          0  951e54f7376e2b2f0915e9e3646c701af4a2fe839385b1...   
130           2          1  

experience  subscribe                                        hashedEmail  \
0             4          1  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1             3          1  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2             3          0  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3             2          1  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4             1          1  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..          ...        ...                                                ...   
191           2          1  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192           3          0  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193           2          0  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194           2          0  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195           4          1  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  prediction_all  
0                 NaN        3.016667  
1                 NaN        0.625000  
2                 NaN        0.625000  
3                 NaN        4.233333  
4                 NaN       19.075000  
..                ...             ...  
191               NaN        4.233333  
192               NaN        0.625000  
193               NaN        4.233333  
194               NaN        4.233333  
195               NaN        3.016667  

[196 rows x 10 columns]

In [31]:

players_plot = (alt.Chart(players_predictions).mark_circle(opacity=0.7).encode(
    x=alt.X("experience").title("Experience").scale(zero=False),
    y=alt.Y("played_hours").title("played_hours").scale(zero=False)
)+
alt.Chart(
    players_predictions,
    title= "K=12"
).mark_line(
    color="black"
).encode(
    x="experience",
    y="prediction_all"
))
players_plot

alt.LayerChart(...)